projects/geocoder-ts/src/lib/cached-geocoder.ts

```diff
@@ -0,0 +1,34 @@
+import { Geocoder } from './models/Geocoder';
+import { Location } from './models/Location';
+
+export class CachedGeocoder implements Geocoder {
+    private addressCache: any = {};
+
+    constructor(private geocoder: Geocoder) { }
+
+    async getLocation(address: string): Promise<Location> {
+        if (this.contains(address)) {
+            return this.lookup(address);
+        }
+
+        const location: Location = await this.geocoder.getLocation(address);
+        this.store(address, location);
+        return location;
+    }
+
+    contains(address: string): boolean {
```

Comment:
This can be greatly simplified by just using a standard Map ... which has a .has function. Less code to test/mess up. Map<string, Location>